### ISSUE TITLE
Enable ANSI color support in Windows console

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/sdk#28614]
With Windows 10, Microsoft added support for ANSI escape codes [1] to the console. With these codes, the user can print text in various colors to the console.

Before these codes can be used, it should be checked if they are actually supported by the terminal (using them in an unsupported terminal yields unreadable output). According to Microsoft, the `GetConsoleMode` API [3] should be used to check if `ENABLE_VIRTUAL_TERMINAL_PROCESSING` is set to determine the terminal's ANSI escape code support.

Unfortunately, I can't find a way to check from Dart if ANSI escape codes are supported on Windows or not. Is there a way? If not, can it be added?

[1] https://en.wikipedia.org/wiki/ANSI_escape_code
[2] https://msdn.microsoft.com/en-us/library/windows/desktop/mt638032(v=vs.85).aspx
[3] https://msdn.microsoft.com/en-us/library/windows/desktop/ms683167(v=vs.85).aspx

**Comment by fsc8000** [comment]
Not sure if there is a good way to expose this directly, but you can detect the Windows version beforehand (using ```ver```), and start Dart with a corresponding environment variable (dart -DSUPPORT_COLOR=true) if the version is Windows 10.

**Comment by goderbauer** [comment]
Even if the user has the right version of Windows, he/she might have `ENABLE_VIRTUAL_TERMINAL_PROCESSING` disabled. So, I kind of do need to check for that (and doing that beforehand from a batch script seems also impossible).

**Comment by fsc8000** [comment]
@floitschG @lrhn  Is there a good place to have this in the io library? Stdout maybe?

**Comment by lrhn** [comment]
Only if it's something we can support on other platforms too, and if it makes sense.
Detecting whether ANSI escape sequences are supported is - tricky. A simple yes/no answer won't cover the actual complexity of which features are supported, and how to use them.

I'm not sure we want to get into the complexity of termcaps, or implement a full termcap-based library (like ncurses) in `dart:io`.

**Comment by goderbauer** [comment]
In addition to just detecting if ANSI support is enabled, it would also be cool if we could enable it from Dart (it's my understanding that it is supposed to be disabled by default on Windows).

**Comment by zanderso** [comment]
I think we need to do something here, and I don't think it has to be perfect. I'm going to add `Platform.ansiSupported` to dart:io, and hard-code it to true when stdout and stdin are connected to a terminal except on Windows, where I'll use the logic from https://github.com/dart-lang/sdk/commit/92b746cd439d94a0ffea7f8bf45669c407acdc96 to decide whether support exists.

https://codereview.chromium.org/2739683005

**Comment by peter-ahe-google** [comment]
Here's one way to detect if colors are available on Unix: https://github.com/dart-lang/sdk/commit/4d302b0adbebb6ee9cdfc99a5e9ecf96c686befe

I'm using the program tput to figure it out, but there should be some equivalent API in C based on termcap.

**Comment by peter-ahe-google** [comment]
When running a program in Emacs, colors aren't available, yet it is a terminal, so I don't think hard-coding Platform.ansiSupported is a good solution. As this feature is implemented now, dart2js and Fasta won't be able to use to detect if colors are supported on Mac and Linux.

**Comment by peter-ahe-google** [comment]
We can't get this to work on Windows 10. We've tried both in Command Prompt and Windows PowerShell.

Are there any other settings we need to enable?

@karlklose

**Comment by zanderso** [comment]
Only very recent Windows 10 builds support both ANSI input and output. Less recent builds support output, but not input. Still earlier Windows 10 builds support neither.

Because of this, I'm moving `ansiSupported` off of `Platform` and adding it as an instance getter on both `Stdin` and `Stdout`.

As for Mac, Linux, etc. I don't think we're going to get a perfect solution here. I don't think we should:
1. Link against ncurses
2. Parse the termcap database by hand
3. Call out to programs on the shell from the standard library

Instead, I'm going to check if the string xterm appears in the TERM environment variable. This should be enough for Terminal on Mac and lots of terminals on Linux including the Gnome terminal:

https://codereview.chromium.org/2753233002

**Comment by goderbauer** [comment]
Here's what my research on the subject turned up:
* Windows 10 Version 1511 (OS build 10586) is the first one to support ANSI output. It was released on November 12, 2015.
* Windows 10 Version 1607 (OS build 14393) is the first one to support ANSI input. It was released on August 2, 2016.

`Platform.ansiSupported` currently only returns `true` on Windows 10 Version 1607 or newer. Once @zanderso's change is in, `Stdout.ansiSupported` should return `true` on Windows 10 Version 1511 or newer and `Stdin.ansiSupported`should return `true` starting on Windows 10 Version 1607 or newer.

**Comment by peter-ahe-google** [comment]
@zanderso I agree, the only option for Linux and Mac appears to link with `ncurses`, and that's not attractive. I hadn't checked the details of how `tput` was implemented. Thank you for fixing this so quickly, I look forward to trying out the colors on Windows next week.

**Comment by nex3** [comment]
It looks like this isn't the complete story. I have a user running Windows 7 who's reporting that Git and Node.js *can* produce colored output, so in principle Dart should be able to as well (see https://github.com/sass/dart-sass/issues/395). Reading [these Windows docs](https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences), it looks like you may be able to call `SetConsoleMode` even on older Windows versions to enable this.

**Comment by mit-mit** [comment]
@goderbauer I have noticed that the `flutter` command on Windows does output color. How does it do that?

**Comment by talamaska** [comment]
Yes there are colors and symbols in Cmd. In Power Shell there is no coloring, and on both the pub outdated command prints checkmark for null-safe packages as question mark, despite that flutter doctor prints everything fine.

**Comment by bsutton** [comment]
@talamaska I have to disagree with your statement on powershell.
Just ran a dart script that outputs red via ansi to a powershell terminal and it works fine. The dart test package also successfully outputs green.

Windows Version	10.0.19042 Build 19042
